### ISSUE TITLE
architecture: eliminate "provider" pattern in favor of vim.ui

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#27119]
### Problem

The old `{runtime}/provider/*` pattern is basically unnecessary since Lua monkey-patching is actually more flexible (and allows stacking, unlike the old pattern). Thus the old "provider" pattern is an unnecessary concept that adds overhead to Nvim development and usage.

### Expected behavior

- Move all the old providers to `vim.ui` (or possibly some other module, but `vim.ui` is probably a fine place for all providers to live).
- Remove support for the old `{runtime}/provider/*` concept, and related documentation.

**Comment by wookayin** [comment]
I agree with the main idea and point made here, i.e. to deprecate all the `*/provider/*` pattern. I think Justin was mainly thinking of clipboard. Existing providers at the moment are rplugins (python, node, perl, ruby) and clipboard:

- Clipboard is a member that belongs to `vim.ui` (as osc52 already does), see `vim.ui.clipboard`
- But rplugins are more like for RPC support and not about UI, so `vim.ui` wouldn't be a good choice. I propose `vim.remote`, `vim.rpc`, etc.

**Comment by justinmk** [comment]
> But rplugins are more like for RPC support and not about UI, so `vim.ui` wouldn't be a good choice. I can suggest: `vim.remote`

I suggested `vim.ui` because introducing something like `vim.providers` just seems pointless because most "providers" really _are_ user-facing interfaces ("UI"). The python/ruby/node providers exist to provide `:pydo`, `:rubydo`, etc, which are directly user-facing features.

**Comment by wookayin** [comment]
> OTOH, they also provide API, so it would be a stretch...

Right. we have `pyeval()` a la `luaeval()`, etc. Even `:python`, `:pydo`, etc. can be seen as for scripting. The concern of "rplugin providers" is too general and wide. If `:pydo`, etc. were about user interface then all other core features (buffers, windows, treesitter, etc.) would also concerning UIs too, after all. Some mid-level utilities and APIs such as notifications, prompts (e.g. `vim.ui.select()`), pop-up (floating windows) as a core util #25514, etc. are directly concerning user interfaces though.

I think a better categorization can focus on their remote/RPC nature (their implementation and protocol) or external Language support for scripting, rather than on their function, i.e., what they are for (remote "plugins", ultimately for better user interface).

**Comment by justinmk** [comment]
Updated to mention `vim.os` instead of `vim.ui`.
